ENH: Change default scoring method to BIC-based for StructureScore

anirudh7371 opened the pull request:
Closes #2302  

This PR updates the default scoring method in `StructureScore.get_scoring_method()` for structure-learning algorithms (e.g., `HillClimbSearch`, `GES`) as follows:

- **Continuous:** Changed default from **Log-Likelihood** → **BIC-based**  
- **Mixed:** Changed default from **Log-Likelihood** → **BIC-based**  
- **Discrete:** Remains **K2** as per the original design  

The change ensures that the default scoring method includes a complexity penalty, leading to more optimal and less dense graphs.

anirudh7371 committed:
feat: Change default structure score to BIC

ankurankan commented:
@anirudh7371 Thanks a lot for the PR. The tests seem to be failing but that is a known issue that we are still trying to figure out.

ankurankan commented:
> Discrete: Remains K2 as per the original design

I would suggest changing this to BIC as well, so that we have a uniform default for all data types.

anirudh7371 commented:
Hi @ankurankan, thanks for the feedback! I’ve updated the default scoring method to BIC for all data types as suggested. However, after this change, some tests like test_all_scores, test_estimate_rand, and test_estimate_titanic are failing because the best DAG structures now have different edge orientations compared to the previous defaults (K2).
Before proceeding, I wanted to check what the preferred approach would be:
Should I update the test expectations to match the new BIC-based defaults? or should we keep K2 as the default for discrete data for now to avoid changing existing behavior? or should I push this as-is and let the maintainers decide on test updates separately?
Please let me know how you’d like me to move forward.

ankurankan committed:
Merge branch 'dev' into fix-default-structure-score

ankurankan commented:
> However, after this change, some tests like test_all_scores, test_estimate_rand, and test_estimate_titanic are failing because the best DAG structures now have different edge orientations compared to the previous defaults (K2).

Could you change these tests to use K2 by passing an extra argument? In that case, we should get the same values as expected in the tests. And then we can add a few more analogous tests using the default scoring method. What do you think? @anirudh7371

anirudh7371 committed:
FIX: Update tests to match new BIC default scorer

anirudh7371 commented:
Hi @ankurankan, thanks for the guidance! I've updated the pull request.

- The original tests now explicitly use the K2 scorer to ensure they pass as before.

- I've added new, separate tests to validate the new BIC default behavior.

All checks are passing now. Please let me know if any other changes are needed.

ankurankan commented:
@anirudh7371 Thanks for the updates. Apart from the linting check failing, everything looks good. I would suggest installing our pre-commit hook for the code to be automatically formatted at each commit: https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md#pre-commit-hooks. 

To fix this PR, you will need to manually run black and commit and push the changes that it makes:
```bash
black pgmpy/tests/test_estimators/test_ExhaustiveSearch.py
```

ankurankan reviewed (changes requested):
Please refer earlier comment.

anirudh7371 committed:
style: Run black on ExhaustiveSearch test file to fix linting

ankurankan reviewed (approved):
@anirudh7371 Looks good. Thanks for the PR.
